### DISCUSSION TITLE
Wrong colors on display with GC9A01 controller

### DISCUSSION THREAD

**Discussion by marsman7** [arendst/Tasmota#19339] [Q&A]
Display : [Waveshare 1.28 inch round LCD modul](https://www.waveshare.com/wiki/1.28inch_LCD_Module)

I tested this display on an ESP32 with the uDisplay driver and the "displaydesc/GC9A01_display.ini" discriptor file and found that the colors are wrong. In this test the order of the primary colors is not RGB (Red Green Blue) instead it is BGR. I have changed in the discriptor file the lines from

```
:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
```

to

```
:R,36
:0,C8,00,00,00
:1,A8,00,00,01
:2,08,00,00,02
:3,68,00,00,03
```

Then the colors were displayed as desired. The basis for this is the color table :

[https://tasmota.github.io/docs/Displays/#color-indices](https://tasmota.github.io/docs/Displays/#color-indices)

Can anyone else confirm this? 

Translated with www.DeepL.com/Translator (free version)

**Comment by shavkatzh** [comment]
Yes. Have same problem for ST7789
The colors under Tasmota work well. Problem found with LVGL settings.

**Reply by s-hadinger** [reply]
What do you mean LVGL settings? Isn't the change in display.ini enough?

**Reply by marsman7** [reply]
yes, the settings above works for me.

**Reply by s-hadinger** [reply]
Could you provide the alternative file to add to the repository?

**Reply by marsman7** [reply]
Yes, i can. I just wanted to make sure that I am not the only one for whom this is the case.
Have you a GC9A01 or a ST7789 display?

**Reply by H3adsho0ot** [reply]
Sadly both versions are not working for my GC9A01 :( DisplayText [x120y120C31] test -> shoud be blue but is yellow with 63488 it shows cyan and with 7 it's  near white. The results were with your modified display ini

**Reply by marsman7** [reply]
There are two ways to specify the color, Color Code and Color Indices.
[https://tasmota.github.io/docs/Displays/#color-codes](https://tasmota.github.io/docs/Displays/#color-codes)

I don't have electronics at hand to test this but you must specify either Cp or Ci not only C. For example

DisplayText [x120y120Ci2] test -> Red Text
or
DisplayText [x120y120Cp63488] test -> Red Text

**Reply by sfromis** [reply]
As I read it, `Cp` does not mean to include the lowercase p, as this indicates the color code, while `Cip` in the parameter description indeed means that the lowercase i is part of the command when using a color index instead of a color code, for certain panels.

**Reply by marsman7** [reply]
In my start post i only test it with Color Indices (Ci).

**Reply by H3adsho0ot** [reply]
DisplayText [x120y120Ci2] test -> Red Text
DisplayText [x120y120C63488] test -> Red Text

both show a cyan "test" on my screen with:
:R,36
:0,C8,00,00,00
:1,A8,00,00,01
:2,08,00,00,02
:3,68,00,00,03

**Reply by H3adsho0ot** [reply]
With default 

:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03

It shows a yellow "test"

**Reply by H3adsho0ot** [reply]
Oh wow a :i was in my display ini .... sorry for that. Without it and with your fixed values it shows the correct colors now!

**Reply by sydknee** [reply]
This seems to have resolved Text but I still have other issues with this.

DisplayMode 1 (useful for testing the connections are correct) I get a white (grey) background and no text.  I'm assuming that the text is the same colour so looking for a way to change that.

The display.ini documentation is confusing for us mere mortals and doesn't really explain much around colours for these modes.  

I would like to rotate the screen by 180 I can tell from the documentation its the :R,36.  I'm assuming 36 is degrees (why not 0) if I put 18 will that work or have I got it totally wrong?

__Just tested the 18 and it worked, strangely the background has now gone black and when I use  DisplayText [x120y120Ci2] test -> Red Text I get red text with a white background but the rest of the background stays black.  Progress!_  _OK spoke to soon, when rotated the colours go back to blue.  Might give the original code a try with  :R,18_

**Reply by sfromis** [reply]
The primary way of selecting your rotation is by using the `DisplayRotate` command, not by modifying `display.ini`.

As the docs say, the values in `:R` are opcodes, which is not about degrees, but about commands to the display controller chip. And the next lines `:1`to `:4` again start with a code, with the 4th parameter selecting what to flip (mirror) and if x/y are to be swapped.

**Reply by sydknee** [reply]
Just trying everything I can, There is something not right about the display.ini for this display and its beyond my abilities.

**Reply by H3adsho0ot** [reply]
Probably you can find some infos here: [https://github.com/arendst/Tasmota/discussions/19487](https://github.com/arendst/Tasmota/discussions/19487)
